### PR TITLE
[flutter_tools] surgically remove outputs from shared directory

### LINKED ISSUE #53667: Write a file containing the last build configuration into the output directory
See https://github.com/flutter/flutter/issues/53546
See https://github.com/flutter/flutter/pull/53601#issuecomment-606737687

As a shorter term solution to multiple builds overlapping in output directories, we can perform a sanity check by writing the last config hash into the output root. If it is different, blow away the folder before writing. Also need to check that nothing else is putting files there.

### LINKED ISSUE #53668: Remove shared output directories from build system
See https://github.com/flutter/flutter/issues/53546
See https://github.com/flutter/flutter/pull/53601#issuecomment-606737687

As a longer term solution, only use the unique build directories - removing the need for https://github.com/flutter/flutter/issues/53667

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Fixes https://github.com/flutter/flutter/issues/53668
Fixes https://github.com/flutter/flutter/issues/53667

When multiple build configurations output to the same directory, the per-target output removal does not work properly, because they do not have knowledge of what the other configuration output. 

To solve this, write a file into the output directory pointing to the last configuration. If this exists and is different from the current configuration, then parse that configurations outputs and surgically remove any that are not in the current output set

### CODE REVIEW

**Comment by jmagman** [comment]
I tried this out locally with `build ios-framework` and I see a `Profile/.last_config` and `Release/.last_config` but no `Debug/.last_config`?  Why would that be?

**Comment by jmagman** [comment]
Also it's such a shame we re-use output directories for build configurations, it's caused us nothing but grief.

**Comment by jonahwilliams** [comment]
> Also it's such a shame we re-use output directories for build configurations, it's caused us nothing but grief.

Yes it is. I would love to change it

**Comment by jonahwilliams** [comment]
The debug ios-framework is going through the build_bundle indirection right? The output directory isn't set right likely. We should fix that to just call assemble

File: packages/flutter_tools/lib/src/build_system/build_system.dart [hunk 1]
```diff
@@ -529,6 +534,57 @@ class BuildSystem {
           ..sort((File a, File b) => a.path.compareTo(b.path)),
     );
   }
+
+  /// Write the configuration of the last build into the output directory and
+  /// remove the previous build's output.
+  ///
+  /// This is used to perform a targeted cleanup of the last output files, if
+  /// these were not already covered by the buil-in cleanup. This is only
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/build_system/build_system.dart:542]: buil->built

**Review by jmagman — COMMENTED** [review]
I checked it out and played with it.  The parts I didn't realize and/or found confusing:
1. The build directory has a unique identifier name and is in `.dart_tool/flutter_build`.  This predates this PR, but I didn't realize it until @jonahwilliams told me even after I checked it out and ran it.  I always thought the buildDir was the `<PROJECT_ROOT>/build` directory?
```dart
  /// Defaults to `{PROJECT_ROOT}/build`. The root of the output directory where
  /// build step intermediates and outputs are written.
  final Directory buildDir;
```
2. The last_config contents are referring to that unique identifier directory name.  Based on the name I thought last_config was going to contain the build mode.
3. outputs.json is in the `.dart_tool/flutter_build` build directory.  Could the `.last_config` contain the JSON file list instead of needing to look it up in the build directory?

File: packages/flutter_tools/lib/src/build_system/build_system.dart [hunk 2]
```diff
@@ -519,6 +519,11 @@ class BuildSystem {
                      path.contains('.dart_tool');
       });
     }
+    trackSharedBuildDirectory(
+      environment, _fileSystem, buildInstance.outputFiles);
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_system/build_system.dart:523]: style nit: if the open paren is on a line without parameters, then the closing paren has to follow suit:
```dart
trackSharedBuildDirectory(
  environment, _fileSystem, buildInstance.outputFiles,
);
```
or
```dart
trackSharedBuildDirectory(environment, _fileSystem, buildInstance.outputFiles);
```

**Review by zanderso — APPROVED** [review]

File: packages/flutter_tools/lib/src/build_system/build_system.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_system/build_system.dart:526]: Done

**Comment by jonahwilliams** [comment]
@jmagman I updated the doc comment on `BUILD_DIR`, it was very much out of date.

2. The .last_config doesn't contain the build mode because it might not be sufficient to distinguish between two builds. 

Originally the build system had privileged configuration like build mode + target platform, but that made it extremely difficult to add new options (like bitcode) since you had to either add a new specific build configuration or ensure that some input somewhere had changed.

3. I could combine the files if you think it makes more sense? I don't feel strongly about it.

**Comment by jonahwilliams** [comment]
@jmagman I updated the name to last build id like we discussed, and expanded the doc comment on trackSharedBuildDirectory

**Review by jmagman — APPROVED** [review]
LGTM, thanks for the changes!

**Comment by jonahwilliams** [comment]
Fails on macOS due to funny business with symlinks

**Pull request merged**